libs/stokesian_dynamics/device_matrix.hpp

```diff
@@ -0,0 +1,658 @@
+        static_assert(std::is_arithmetic<T>::value,
+                      "Data type of device_matrix must be arithmetic for "
+                      "arithmetic operations");
+        return m_rows == B.m_rows && m_cols == B.m_cols &&
+               thrust_wrapper::equal(Policy::par(), data(), data() + size(), 
+                      B.data(), internal::almostEqual<value_type>{epsilon});
+    }
+
+    /// \}
+
+    /// \defgroup linalg Linear algebra
+    /// \{
+
+    /// Compute the transpose.
+    device_matrix transpose() const {
+        static_assert(std::is_same<T, double>::value,
+                      "Data type of device_matrix must be floating point for "
+                      "BLAS/LAPACK operations");
+        device_matrix C(m_cols, m_rows);
+        internal::cublas<Policy, value_type>::geam(
+            thrust_wrapper::raw_pointer_cast(data()),
+            thrust_wrapper::raw_pointer_cast(C.data()), m_rows, m_cols);
+        return C;
+    }
+
+    /// Compute the inverse and the Cholesky decomposition.
+    thrust_wrapper::tuple<device_matrix, device_matrix> inverse_and_cholesky() const {
+        static_assert(std::is_same<T, double>::value,
+                      "Data type of device_matrix must be floating point for "
+                      "BLAS/LAPACK operations");
+        assert(m_rows == m_cols);
+        device_matrix A = *this;
+        device_matrix B = device_matrix::Identity(m_rows, m_cols);
+
+        // debugging stuff
+//        double *raw_data = thrust_wrapper::raw_pointer_cast(B.data());
+//        printf("A.data: \n");
+//        for (int i=0; i<m_rows; i++) {
+//          for (int j=0; j<m_cols; j++) {
+//            printf("%f ",*raw_data);
+//            raw_data++;
+//          }
+//          printf("\n");
+//        }
+        
+
+        internal::cusolver<Policy, double>::potrf(
+            thrust_wrapper::raw_pointer_cast(A.data()),
+            thrust_wrapper::raw_pointer_cast(B.data()), m_rows);
+
+        return thrust_wrapper::make_tuple(B, A);
+    }
+
+    /// Compute the inverse.
+    device_matrix inverse() const {
+        return thrust_wrapper::get<0>(inverse_and_cholesky());
+    }
+
+    /// \}
+
+    /// \defgroup generators Convenience generator functions
+    /// \{
+
+    /// Generate an identity matrix with size \p rows x \p cols.
+    static device_matrix Identity(size_type rows, size_type cols) {
+        static_assert(std::is_same<T, double>::value,
+                      "Data type of device_matrix must be floating point for "
+                      "BLAS/LAPACK operations");
+        device_matrix I(rows, cols);
+        thrust_wrapper::tabulate(Policy::par(), I.data(), I.data() + I.size(),
+                                 internal::IdentityGenerator<value_type>{rows});
+        return I;
+    }
+
+    /// \}
+};
+
+
+/** Simpler version of device_matrix that doesn't provide the arithmetic
+ *  operations. Can run on a THRUST DEVICE. 
```

Comment:
The comment states that the access to sub-blocks of a matrix is not used in the SD library. Why is the matrix view class needed at all, then?

Reply:
It is used as a read-only reference to other device_matrix objects and cannot be trivially replaced AFAIK.